packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts

```diff
@@ -473,6 +473,53 @@ export default class UserStorageController extends BaseController<
     );
   }
 
+  #isLocked = false;
+
+  readonly #pendingOperations: (() => Promise<unknown> | unknown)[] = [];
+
+  /**
+   * Locks the controller, preventing operations that are subsequently called with `executeWithLockHandling` from executing.
+   * These operations will be queued until the controller is unlocked.
+   */
+  lock() {
+    this.#isLocked = true;
+  }
+
+  /**
+   * Unlocks the controller and executes all pending operations that were queued using the `executeWithLockHandling` method while the controller was locked.
+   * This method will also clear the pending operations queue.
+   * If any of the operations fail, the error will be propagated to the caller. The following operations will not be executed and the queue will be cleared.
```

Comment:
Forwarding errors to the caller here creates a short circuit between caller stacks.
An enqueued operation that fails will look like a failure of the `unlock()` call, or like a failure of another caller if my suggestion from below is applied.